servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultAsyncContextProvider.java

```diff
@@ -104,165 +104,115 @@ public final CompletableSource.Subscriber wrapCancellable(final CompletableSourc
+        return new ContextPreservingCompletableSubscriber(subscriber, null, context);
     }
 
     @Override
     public final CompletableSource.Subscriber wrapCompletableSubscriberAndCancellable(
             final CompletableSource.Subscriber subscriber, final CapturedContext context) {
         if (subscriber instanceof ContextPreservingCompletableSubscriber) {
             final ContextPreservingCompletableSubscriber s = (ContextPreservingCompletableSubscriber) subscriber;
-            if (s.capturedContext == context) {
-                return subscriber instanceof ContextPreservingCompletableSubscriberAndCancellable ? subscriber :
-                        new ContextPreservingCompletableSubscriberAndCancellable(s.subscriber, context);
-            }
-        } else if (subscriber instanceof ContextPreservingCancellableCompletableSubscriber) {
-            final ContextPreservingCancellableCompletableSubscriber s =
-                    (ContextPreservingCancellableCompletableSubscriber) subscriber;
-            if (s.capturedContext == context) {
-                return new ContextPreservingCompletableSubscriberAndCancellable(s.subscriber, context);
+            if (s.cancellableCapturedContext == context && s.subscriberCapturedContext != null) {
+                // Subscriber already wrapped.
+                return subscriber;
             }
+            CapturedContext subscriberContext = s.subscriberCapturedContext == null ?
+                    context : s.subscriberCapturedContext;
+            return new ContextPreservingCompletableSubscriber(s.subscriber, context, subscriberContext);
         }
-        return new ContextPreservingCompletableSubscriberAndCancellable(subscriber, context);
+        return new ContextPreservingCompletableSubscriber(subscriber, context, context);
     }
 
     @Override
     public final <T> SingleSource.Subscriber<T> wrapCancellable(final SingleSource.Subscriber<T> subscriber,
                                                                 final CapturedContext context) {
         if (subscriber instanceof ContextPreservingSingleSubscriber) {
             final ContextPreservingSingleSubscriber<T> s = (ContextPreservingSingleSubscriber<T>) subscriber;
-            if (s.capturedContext == context) {
-                return subscriber instanceof ContextPreservingSingleSubscriberAndCancellable ? subscriber :
-                        new ContextPreservingSingleSubscriberAndCancellable<>(s.subscriber, context);
-            }
-        } else if (subscriber instanceof ContextPreservingCancellableSingleSubscriber &&
-                ((ContextPreservingCancellableSingleSubscriber<T>) subscriber).capturedContext == context) {
-            // no need to check for instanceof ContextPreservingSingleSubscriberAndCancellable, because
-            // it extends from ContextPreservingSingleSubscriber.
-            return subscriber;
+            return s.cancellableCapturedContext == context ? subscriber :
```

Comment:
Comment clarifying equality check will be helpful for ppl looking at this code later